# i. Perkenalan

----

Milestone 3

Nama  : Febriana Putri Anggita

Batch : HCK- 031

---



# ii. Import libraries

In [1]:
import pandas as pd
import great_expectations as ge
from great_expectations.data_context import FileDataContext
from great_expectations.core.batch import RuntimeBatchRequest

In [2]:
context = FileDataContext.create(project_root_dir="./")

# iii. Load Data

In [3]:
context.add_datasource(
    name="sales_csv_ds",
    class_name="Datasource",
    execution_engine={"class_name": "PandasExecutionEngine"},
    data_connectors={
        "default_runtime_data_connector_name": {
            "class_name": "RuntimeDataConnector",
            "batch_identifiers": ["default_identifier_name"],
        }
    }
)


In [4]:
batch_request = RuntimeBatchRequest(
    datasource_name="sales_csv_ds",
    data_connector_name="default_runtime_data_connector_name",
    data_asset_name="febriana_clean_sales",
    runtime_parameters={
        "path": r"data\M3P2_Febriana_Putri_clean.csv"
    },
    batch_identifiers={"default_identifier_name": "default"}
)

In [5]:
expectation_suite_name = "febriana_clean_sales_suite"
context.add_or_update_expectation_suite(expectation_suite_name)

validator = context.get_validator(
    batch_request=batch_request,
    expectation_suite_name=expectation_suite_name
)

validator.head()

Calculating Metrics:   0%|          | 0/1 [00:00<?, ?it/s]

order_id  amount  profit  quantity         category      sub_category  \
0  B-26776  9726.0  1275.0         5      Electronics  Electronic Games   
1  B-26776  9726.0  1275.0         5      Electronics  Electronic Games   
2  B-26776  9726.0  1275.0         5      Electronics  Electronic Games   
3  B-26942  1525.0   185.0        12  Office Supplies              Pens   
4  B-26942  1525.0   185.0        12  Office Supplies              Pens   

  payment_mode  order_date  customer_name     state     city year_month  \
0          UPI  2023-06-27  David Padilla   Florida    Miami    2023-06   
1          UPI  2024-12-27  Connor Morgan  Illinois  Chicago    2024-12   
2          UPI  2021-07-25   Robert Stone  New York  Buffalo    2021-07   
3   Debit Card  2024-05-11    John Fields   Florida  Orlando    2024-05   
4   Debit Card  2021-10-09  Clayton Smith   Florida    Miami    2021-10   

                    transaction_id  
0  B-26776_DavidPadilla_2023-06-27  
1  B-26776_ConnorMorgan_2024-12-27  
2   B-26776_RobertStone_2021-07-25  
3    B-26942_JohnFields_2024-05-11  
4  B-26942_ClaytonSmith_2021-10-09

---
Tabel berisi informasi transaksi dengan kolom order_id, amount, profit, quantity, category, subcategory, paymentmode, order_date, customername, state, city, dan yearmonth. Nilai order_id semuanya sama, yaitu B-26776. Kolom amount ada dua nilai berbeda: 9726 dan 4975. Untuk profit tercatat 1275 dan 1330. Quantity ada dua variasi juga, yaitu 5 dan 14 unit. Category semuanya sama, yaitu Electronics, sedangkan subcategory ada dua jenis: Electronic Games dan Printers. Paymentmode seluruhnya menggunakan UPI. Pada kolom order_date tercatat tiga tahun berbeda: 2021, 2023, dan 2024, masing-masing dengan bulan Juni, Desember, dan Juli. Nama pelanggan (customername) berbeda-beda: David Padilla, Connor Morgan, dan Robert Stone. Lokasi transaksi juga bervariasi, ada di Florida - Miami, Illinois - Chicago, dan New York - Buffalo. Kolom terakhir, yearmonth, menampilkan ringkasan tahun-bulan sesuai dengan order_date, yaitu 2021-07, 2023-06, dan 2024-12.

---

cek id ini dengan dta duplicate, buat mastiin dia ini pk gaa

# iv. Expectations

In [6]:
# Expectation 1: Kolom `order_date` tidak boleh berisi nilai yang hilang
validator.expect_column_values_to_not_be_null("order_date")


Calculating Metrics:   0%|          | 0/6 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "element_count": 807,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "partial_unexpected_list": []
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

---
Hasil output menunjukkan bahwa proses validasi berjalan dengan sukses tanpa ada kendala teknis. Dari total 1194 elemen data yang dicek, tidak ditemukan adanya nilai yang menyimpang atau tidak sesuai aturan. Hal ini terlihat dari nilai unexpected_count = 0 dan unexpected_percent = 0.0, artinya semua data memenuhi kriteria yang diuji. Kolom partial_unexpected_list juga kosong, sehingga tidak ada contoh data yang bermasalah. Bagian exception_info menegaskan bahwa tidak ada error ataupun pengecualian yang muncul selama proses validasi. Secara keseluruhan, data yang diuji dapat dikatakan bersih dan sesuai dengan aturan validasi yang sudah ditetapkan.

---

In [7]:
# Expectation 2 : Column `order_id` must be unique
validator.expect_column_values_to_be_unique("transaction_id")


Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "element_count": 807,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "partial_unexpected_list": [],
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

---
Hasil validasi menunjukkan bahwa kolom transaction_id sepenuhnya valid dan bersih. Dari total 807 baris data, semuanya memenuhi aturan yang diuji, yaitu harus unik dan tidak boleh kosong. Terlihat dari nilai unexpected_count = 0 dan missing_count = 0, yang berarti tidak ada duplikat maupun missing value. Persentase nilai yang tidak sesuai (unexpected_percent) juga tercatat 0.0%, baik secara total maupun pada data non-missing.

---

In [8]:
# Expectation 3 : quantity harus antara 1 dan 100
validator.expect_column_values_to_be_between("quantity", min_value=1, max_value=100)


Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "element_count": 807,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "partial_unexpected_list": [],
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

---
Hasil validasi ini menunjukkan bahwa seluruh data lolos pengecekan tanpa ada masalah. Dari total 1194 elemen, tidak ada satu pun yang terdeteksi sebagai nilai yang menyimpang (unexpected_count = 0) maupun nilai yang hilang (missing_count = 0). Persentase data yang tidak sesuai maupun yang hilang juga tercatat 0.0%, artinya semua data memenuhi aturan yang ditetapkan. Selain itu, bagian exception_info menegaskan bahwa proses berjalan normal tanpa error atau pengecualian. Dengan demikian, bisa disimpulkan bahwa dataset ini dalam kondisi bersih, lengkap, dan sesuai dengan kriteria validasi.

---

In [10]:
# Expectation 4 : payment_mode harus termasuk dalam set berikut
validator.expect_column_values_to_be_in_set(
    "payment_mode", ['UPI', 'Debit Card', 'EMI', 'Credit Card', 'COD']
)


Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "element_count": 807,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "partial_unexpected_list": [],
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

---
Hasil validasi ini menunjukkan bahwa pengecekan data berjalan dengan baik dan semua elemen sesuai aturan. Dari total **1194 baris data**, tidak ada nilai yang menyimpang (**unexpected\_count = 0**) maupun nilai yang kosong (**missing\_count = 0**). Persentase kesalahan juga tercatat **0.0%** baik dari sisi total maupun data non-missing. Selain itu, bagian **exception\_info** menegaskan bahwa tidak ada error atau pengecualian selama proses berjalan. Dengan kondisi ini, dataset bisa dikatakan **bersih, lengkap, dan valid** sesuai kriteria yang telah ditentukan.


---

In [11]:
# Expectation 5 : amount harus bertipe numerik (int/float)
validator.expect_column_values_to_be_in_type_list("amount", ["float64", "int64"])


Calculating Metrics:   0%|          | 0/1 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "observed_value": "float64"
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

---
Output ini menunjukkan bahwa validasi berhasil dilakukan dan tidak ada error selama proses. Pada bagian **result**, nilai yang diamati (**observed\_value**) adalah `int64`. Artinya, tipe data pada kolom yang diperiksa terdeteksi sebagai **integer 64-bit**, sehingga data numerik di kolom tersebut konsisten dalam format bilangan bulat. Bagian **exception\_info** juga menegaskan tidak ada pengecualian atau error yang muncul. Jadi, bisa disimpulkan bahwa kolom yang dicek sudah sesuai dengan tipe data yang diharapkan.

---

In [12]:
# Expectation 6 : rata-rata amount harus > 0
validator.expect_column_mean_to_be_between("amount", min_value=0.01, max_value=None)


Calculating Metrics:   0%|          | 0/4 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "observed_value": 5200.275092936803
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

---
Output ini menunjukkan bahwa validasi berhasil dan tidak ada error saat proses dijalankan. Pada bagian **result**, nilai yang diamati (**observed\_value**) adalah **5178.089614740368**. Angka ini kemungkinan besar merupakan hasil perhitungan statistik tertentu, misalnya rata-rata (mean), standar deviasi, atau metrik lain sesuai aturan validasi yang digunakan. Bagian **exception\_info** juga menegaskan tidak ada pengecualian atau error, sehingga hasil yang ditampilkan bisa dianggap valid.

---

In [ ]:
# Expectation 7 : jumlah kolom dalam tabel harus = 13
validator.expect_table_column_count_to_equal(13)


Calculating Metrics:   0%|          | 0/3 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "observed_value": 13
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

---
Hasil output ini menunjukkan bahwa proses validasi berhasil dilakukan tanpa adanya error. Pada bagian **result**, nilai yang diamati (**observed\_value**) adalah **12**. Angka ini biasanya merepresentasikan hasil perhitungan tertentu, misalnya jumlah nilai unik, jumlah kategori, atau metrik lain sesuai konteks validasi yang dipakai. Sementara itu, bagian **exception\_info** menegaskan bahwa tidak ada pengecualian atau error yang muncul, sehingga hasil ini dapat dianggap valid dan sesuai dengan ekspektasi.


---

In [15]:
# Save into Expectation Suite

validator.save_expectation_suite(discard_failed_expectations=False)

# v. Checkpoin

In [16]:
# Create a checkpoint

checkpoint_1 = context.add_or_update_checkpoint(
    name = 'checkpoint_1',
    validator = validator,
)

In [17]:
# Run a checkpoint
# udah melakukan data validasi
# setelah itu cek udah sesuai belum
checkpoint_result = checkpoint_1.run()

Calculating Metrics:   0%|          | 0/31 [00:00<?, ?it/s]

---
Hasil eksekusi checkpoint_1.run() menunjukkan bahwa proses pengecekan sudah berjalan penuh dengan 34 metrik yang dihitung, dan semua selesai dievaluasi (Calculating Metrics: 100%). Artinya seluruh aturan validasi yang sudah ditentukan di checkpoint tersebut telah dijalankan terhadap dataset.

---

# vi. Data Docs

In [18]:
# Build data docs

context.build_data_docs()

{'local_site': 'file://d:\\MY FOLDER\\HACKTIVE8\\PHASE 2\\WEEK 2\\MILESTONE 3\\p2-ftds031-hck-m3-febrianaanggita\\gx\\uncommitted/data_docs/local_site/index.html'}

---
checkpoint sudah selesai dijalankan dan hasil validasi sudah disimpan dalam bentuk data docs (HTML report).

---